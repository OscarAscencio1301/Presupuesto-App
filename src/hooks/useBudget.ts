import { useDispatch, useSelector } from "react-redux"
import { Budget, Expense, Slice } from "../interface/budget"
import { active, add, calculate, changeModal, clean, defineBudget, deleteB, filterExpense, resetValues, update } from "../store/budget/budgetSlice"


export const useBudget = () => {
    const { expenses, expenseActive, search, expensesFilter, budget, available, spent, porcent, isValidBudget, isOpenModal } = useSelector<Slice, Budget>(state => state.budget)
    const dispatch = useDispatch()


    const defineBudgetAct = (budget: number) => {
        dispatch(defineBudget(budget))
    }
    const changeModalAct = () => {
        dispatch(changeModal())
    }

    const addExpenseAct = (expense: Expense) => {
        dispatch(add(expense))
    }
    const activeExpenseAct = (expense: Expense) => {
        dispatch(active(expense))
    }
    const updateExpenseAct = (expense: Expense) => {
        dispatch(update(expense))
    }
    const deleteExpenseAct = (id: number) => {
        dispatch(deleteB(id))
    }
    const cleanExpenseAct = () => {
        dispatch(clean())
    }

    const calculateAct = () => {
        dispatch(calculate())
    }

    const filterExpensesAct = (category:string) => {
        dispatch(filterExpense(category))
    }

    const resetAct = () => {
        dispatch(resetValues())
    }

    return {
        expenses,
        expenseActive,
        expensesFilter,
        search,
        porcent,
        available,
        spent,
        budget,
        isOpenModal,
        isValidBudget,
        defineBudgetAct,
        changeModalAct,
        addExpenseAct,
        activeExpenseAct,
        updateExpenseAct,
        deleteExpenseAct,
        cleanExpenseAct,
        calculateAct,
        filterExpensesAct,
        resetAct
    }
}
