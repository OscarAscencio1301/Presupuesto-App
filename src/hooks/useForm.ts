import { ChangeEvent, useState } from "react"


export const useForm = <T>(initialState: T) => {

    const [form, setform] = useState(initialState)

    const changeEvent = ({ target: { name, value } }: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
        setform({
            ...form,
            [name]: value
        })
    }

    const reset = (newValues = initialState) => {
        setform(newValues)
    }

    return {
        ...form,
        form,
        setform,
        changeEvent,
        reset
    }
}
