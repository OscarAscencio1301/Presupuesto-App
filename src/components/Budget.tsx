import { ChangeEvent, FormEvent, useState } from "react"
import { useBudget } from "../hooks/useBudget"

export const Budget = () => {
    const { defineBudgetAct } = useBudget()
    const [budgetPrev, setbudgetPrev] = useState(0)

    const handleBudgetChange = ({target:{value}}: ChangeEvent<HTMLInputElement>) => {
        const inputBudget = value;
        setbudgetPrev(Number(inputBudget));
    };


    const sendBudget = (e: FormEvent) => {
        e.preventDefault()
        if (isNaN(budgetPrev) || budgetPrev <= 0) return
        defineBudgetAct(Number(budgetPrev))
    }




    return (
        <form className="bg-white container mx-auto max-w-screen-lg rounded-lg gap-4 flex-col flex justify-center p-5 lg:p-10 shadow-md" onSubmit={sendBudget}>
            <h3 className="text-green-700 text-3xl font-medium text-center">Definir Presupuesto</h3>
            <input
                type="number"
                name="budget"
                placeholder="Añade tu presupuesto"
                className="text-xl border p-2 text-black outline-none placeholder:text-center"
                value={budgetPrev}
                onChange={handleBudgetChange} // Utilizamos la función handleBudgetChange
            />
            <button type="submit" className="bg-green-700 py-2 px-10 rounded-md hover:bg-green-900">Presupuesto</button>
        </form>
    )
}
