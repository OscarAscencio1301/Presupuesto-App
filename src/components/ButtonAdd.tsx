import { useBudget } from "../hooks/useBudget"
import imageBudgetNew from '../assets/nuevo-gasto.svg'

export const ButtonAdd = () => {
    const { changeModalAct } = useBudget()
    return (
        <div className="fixed bottom-1 right-1 w-16 max-w-md m-3 cursor-pointer" onClick={changeModalAct}>
            <img src={imageBudgetNew} alt="New Budget" />
        </div>
    )
}
