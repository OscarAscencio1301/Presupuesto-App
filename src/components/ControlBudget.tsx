import { useEffect, useMemo } from "react"
import { useBudget } from "../hooks/useBudget"

import { CircularProgressbar, buildStyles } from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';

export const ControlBudget = () => {
    const { budget, available, spent, calculateAct, expenses ,porcent } = useBudget()

    const [formatBudget, formatavailable, formatspent] = useMemo(() => {
        return [
            budget.toLocaleString('en-US', { style: 'currency', currency: 'USD' }),
            available.toLocaleString('en-US', { style: 'currency', currency: 'USD' }),
            spent.toLocaleString('en-US', { style: 'currency', currency: 'USD' }),
        ]
    }, [budget, available, spent])


    useEffect(() => {
        calculateAct()
    }, [expenses])


    return (
        <div className="bg-white container mx-auto max-w-screen-lg rounded-lg gap-4 flex-col flex justify-center items-center p-5 lg:p-10 shadow-md">
            <div className="">
                <CircularProgressbar value={porcent} text={`${porcent}% Disponible`} styles={buildStyles({
                    pathColor: '#16A34A',
                    textColor: '#16A34A',
                    textSize: '10px'
                })}/>;
            </div>
            <div className="flex-1 text-center">
                <p className="text-black text-2xl"><span className="text-green-600">Presupuesto: </span>{formatBudget}</p>
                <p className="text-black text-2xl"><span className="text-green-600">Disponible: </span>{formatavailable}</p>
                <p className="text-black text-2xl"><span className="text-green-600">Gastada: </span>{formatspent}</p>
            </div>
        </div>
    )
}
