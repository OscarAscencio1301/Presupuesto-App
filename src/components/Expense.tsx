import { useMemo } from "react"
import { LeadingActions, SwipeableList, SwipeAction, SwipeableListItem, TrailingActions } from "react-swipeable-list"
import { Expense } from "../interface/budget"
import { convertDate } from '../helpers/ConvertDate'
import imageBudget from '../assets/icono_ahorro.svg'
import imageHouse from '../assets/icono_casa.svg'
import imageFoot from '../assets/icono_comida.svg'
import imageExpense from '../assets/icono_gastos.svg'
import imageHospital from '../assets/icono_salud.svg'
import imageSubs from '../assets/icono_suscripciones.svg'
import imageOcio from '../assets/icono_ocio.svg'
import "react-swipeable-list/dist/styles.css"
import { useBudget } from "../hooks/useBudget"

interface Icons {
    [key: string]: string;
}

const icons: Icons = {
    ahorro: imageBudget,
    comida: imageFoot,
    casa: imageHouse,
    gastos: imageExpense,
    ocio: imageOcio,
    salud: imageHospital,
    subscripciones: imageSubs
}

export const SingleExpense = ({ category, cost, expense, id }: Expense) => {

    const date = useMemo(() => convertDate(Number(id)), [id])
    const { deleteExpenseAct, activeExpenseAct, changeModalAct } = useBudget()

    const functionLeft = () => (
        <LeadingActions>
            <SwipeAction onClick={() => {
                changeModalAct()
                activeExpenseAct({ category, cost, expense, id })
            }}>
                Editar
            </SwipeAction>
        </LeadingActions>
    )
    const functionRight = () => (
        <TrailingActions>
            <SwipeAction onClick={() => deleteExpenseAct(id || 0)}>
                Eliminar
            </SwipeAction>
        </TrailingActions>
    )

    return (
        <SwipeableList>
            <SwipeableListItem
                leadingActions={functionLeft()}
                trailingActions={functionRight()}
            >
                <div className="shadow flex-1 flex items-center justify-between gap-5 mb-5 p-5">
                    <img src={icons[category]} alt="Category Image" className="w-20" />
                    <div className="text-3xl text-black font-bold">
                        <div className="mb-1">
                            <p className="text-gray-500 text-xl uppercase">{category}</p>
                            <p className="text-gray-800 text-2xl">{expense}</p>
                            <p className="text-lg text-gray-700">{date}</p>
                        </div>
                    </div>
                    <div className="">
                        <p className="text-2xl text-black font-bold">${cost}</p>
                    </div>
                </div>
            </SwipeableListItem>
        </SwipeableList>
    )
}
