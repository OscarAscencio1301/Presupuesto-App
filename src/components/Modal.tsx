import imageClose from '../assets/cerrar.svg'
import { useBudget } from '../hooks/useBudget'

export const Modal = () => {

    const {isOpenModal, changeModalAct} = useBudget()

  return (
    <div className={`top-0 absolute z-10 h-screen w-screen bg-black bg-opacity-90 ${isOpenModal ? 'block' : 'hidden'}`}>
        <div className="w-10 absolute right-10 top-5 cursor-pointer" onClick={changeModalAct}>
            <img src={imageClose} alt="Close Modal" />
        </div>
    </div>
  )
}
