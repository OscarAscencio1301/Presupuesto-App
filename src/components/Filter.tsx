import { useEffect } from "react"
import { useBudget } from "../hooks/useBudget"
import { useForm } from "../hooks/useForm"

export const Filter = () => {
  const {filterExpensesAct} = useBudget()
  const {search, changeEvent} = useForm({
    search: ''
  })

  useEffect(() => {
    filterExpensesAct(search)
  }, [search])
  
  

  return (
    <div className="shadow py-5 px-8">
      <form>
        <div className="flex items-center gap-10">
          <label className="text-3xl font-bold text-gray-700">Filtrar gastos</label>
          <select className="flex-1 p-3 border-none rounded-sm text-center" name="search" value={search} onChange={changeEvent}>
            <option value="">-- Seleccione --</option>
            <option value="ahorro">Ahorro</option>
            <option value="comida">Comida</option>
            <option value="casa">Casa</option>
            <option value="gastos">Gastos Varioa</option>
            <option value="ocio">Ocio</option>
            <option value="salud">Salud</option>
            <option value="subscripciones">Subscripciones</option>

          </select>
        </div>
      </form>
    </div>
  )
}
