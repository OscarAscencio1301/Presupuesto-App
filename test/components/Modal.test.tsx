import { fireEvent, render, screen } from "@testing-library/react";
import { Modal } from "../../src/components/Modal";
import { useBudget } from "../../src/hooks/useBudget";
import { useForm } from "../../src/hooks/useForm";
jest.mock('../../src/hooks/useBudget');
jest.mock('../../src/hooks/useForm');

describe('Testing Component Modal', () => {

    const addElementMock = jest.fn()

    beforeEach(() => jest.clearAllMocks())

    test('Snapshot ', () => {

        (useBudget as jest.Mock).mockReturnValue({
            isOpenModal: true,
            changeModalAct: jest.fn(),
            addExpenseAct: jest.fn(),
            expenseActive: jest.fn(),
            updateExpenseAct: jest.fn(),
            cleanExpenseAct: jest.fn(),
        });

        (useForm as jest.Mock).mockReturnValue({
            setform: jest.fn(),
        });

        const { container } = render(<Modal />)

        expect(container).toMatchSnapshot()
    });

    test('Initial Values', () => {

        (useBudget as jest.Mock).mockReturnValue({
            isOpenModal: true,
            changeModalAct: jest.fn(),
            addExpenseAct: jest.fn(),
            expenseActive: jest.fn(),
            updateExpenseAct: jest.fn(),
            cleanExpenseAct: jest.fn(),
        });

        (useForm as jest.Mock).mockReturnValue({
            setform: jest.fn(),
        });

        render(<Modal />)
        expect(screen.getByText('Nuevo Gasto')).toBeTruthy()


    });

    test('Change Values Form', () => {

        (useBudget as jest.Mock).mockReturnValue({
            isOpenModal: true,
            changeModalAct: jest.fn(),
            addExpenseAct: addElementMock,
            expenseActive: null,
            updateExpenseAct: jest.fn(),
            cleanExpenseAct: jest.fn(),
        });

        (useForm as jest.Mock).mockReturnValue({
            setform: jest.fn(),
        });

        render(<Modal />)
        const nameExpense = screen.getByLabelText<HTMLInputElement>('expense');
        const cost = screen.getByLabelText<HTMLInputElement>('cost');
        const category = screen.getByLabelText<HTMLSelectElement>('category')

        fireEvent.input(nameExpense, { target: { value: 'Nuevo Gasto' } })
        fireEvent.input(cost, { target: { value: '2000' } })
        fireEvent.select(category, { target: { value: 'salud' } });

        expect(nameExpense.value).toBe('Nuevo Gasto')
        expect(cost.value).toBe('2000')
        expect(category.value).toEqual('salud')


    });

    test('Submit Form', () => {

        (useBudget as jest.Mock).mockReturnValue({
            isOpenModal: true,
            changeModalAct: jest.fn(),
            addExpenseAct: addElementMock,
            expenseActive: null,
            updateExpenseAct: jest.fn(),
            cleanExpenseAct: jest.fn(),
        });

        (useForm as jest.Mock).mockReturnValue({
            setform: jest.fn(),
            reset: jest.fn()
        });

        render(<Modal />)
        const nameExpense = screen.getByLabelText<HTMLInputElement>('expense');
        const cost = screen.getByLabelText<HTMLInputElement>('cost');
        const category = screen.getByLabelText<HTMLSelectElement>('category')
        const form = screen.getByLabelText('form')

        fireEvent.input(nameExpense, { target: { value: 'Nuevo Gasto' } })
        fireEvent.input(cost, { target: { value: '2000' } })
        fireEvent.select(category, { target: { value: 'salud' } });
        fireEvent.submit(form);

        expect(addElementMock).toBeCalled()



    });
})