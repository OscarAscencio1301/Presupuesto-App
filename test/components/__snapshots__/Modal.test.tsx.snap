// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Component Modal Snapshot  1`] = `
<div>
  <div
    class="flex justify-center items-center top-0 absolute z-10 h-screen w-screen bg-black bg-opacity-90 block"
  >
    <div
      class="w-10 absolute right-3 top-5 cursor-pointer"
    >
      <img
        alt="Close Modal"
      />
    </div>
    <form
      aria-label="form"
      class="flex flex-col w-[400px] p-5 gap-10 justify-center h-screen"
    >
      <legend>
        Nuevo Gasto
      </legend>
      <div
        class="text-white flex flex-col gap-2 "
      >
        <label
          class="text-3xl"
          for=""
        >
          Nombre Gasto
        </label>
        <input
          aria-label="expense"
          class="bg-white border rounded p-2 text-black"
          name="expense"
          placeholder="Añade nombre al gasto"
          type="text"
          value=""
        />
      </div>
      <div
        class="text-white flex flex-col gap-2 "
      >
        <label
          class="text-3xl"
          for=""
        >
          Gasto
        </label>
        <input
          aria-label="cost"
          class="bg-white border rounded p-2 text-black"
          name="cost"
          placeholder="Añade el gasto"
          type="number"
          value=""
        />
      </div>
      <div
        class="text-white flex flex-col gap-2 "
      >
        <label
          class="text-3xl"
          for=""
        >
          Categoría
        </label>
        <select
          aria-label="category"
          class="bg-white border rounded p-2 text-black"
          name="category"
        >
          <option
            value=""
          >
            -- Seleccione --
          </option>
          <option
            value="ahorro"
          >
            Ahorro
          </option>
          <option
            value="comida"
          >
            Comida
          </option>
          <option
            value="casa"
          >
            Casa
          </option>
          <option
            value="gastos"
          >
            Gastos Varioa
          </option>
          <option
            value="ocio"
          >
            Ocio
          </option>
          <option
            value="salud"
          >
            Salud
          </option>
          <option
            value="subscripciones"
          >
            Subscripciones
          </option>
        </select>
      </div>
      <button
        class="bg-green-700 text-white p-2 rounded-sm hover:bg-green-900"
        type="submit"
      >
        Agregar gasto
      </button>
    </form>
  </div>
</div>
`;
